### PR TITLE
Add configurable fixed partitioning scheme

### LINKED ISSUE #6266: Support configuration of rack-aware partition distribution
**Is your feature request related to a problem? Please describe.**

In a multi-datacenter setup it is required to control which partitions are located in which data center. It also is advantageous if a user can indicate which brokers should be priority leader for a partition, for example to indicate the preferred leader balance between the data centers.

**Describe the solution you'd like**

In the cluster configuration I can specify a manual mapping of partitions hosted by the broker. This will be used when defining the partition membership on startup, and ignore the default partition layout over the cluster. Additionally a user can specify a priority for a leader to be leader of a specific partition. This priority should be used to influence a heuristic in the leader election process which prefers the prioritized broker. 

**Describe alternatives you've considered**

- Setup the cluster in a way that the current partition layout leads to an even distribution
- Restart the cluster until the expected leader balancing happened

**Additional context**

If the user chooses to manually configure the partition layout, its there responsibility to ensure that partitions have the correct number of replicas and are evenly distributed. The broker will only perform sanity checks possible from a single broker configuration.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR introduces a new configurable partitioning scheme which lets users manually distribute partitions across a static set of nodes. Additionally, if priority election is used, it lets users assign priorities to these nodes as well.

A valid configuration consists of:

- All partitions with IDs between 1 and `partitionsCount` are configured explicitly
- All partitions configure exactly `replicationFactor` nodes
- All nodes configured for a partition have a valid node ID, i.e. between 0 and `clusterSize - 1`.
- If priority election is enabled, then all priorities configured for a partition are different

If any of these conditions are not met, then the broker will fail to start. Note that if priority election is enabled, the target priority for a partition will be whatever was the highest configured priority.

## Related issues

closes #6266 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by npepinpe** [comment]
@deepthidevaki - I initially had a very lightweight model for the distributor, where it would just take in a map of maps, and figure out the distribution. You can see this [commit](https://github.com/camunda-cloud/zeebe/commit/3d360cf3d6f61041a8d8c221c102078b214caee3) where I added a more explicit intermediate model. I like this more, but it's definitely more verbose. The main reason why I like this approach more:

- It's more explicit in terms of what are the map keys and values, what they mean, etc., then using some basic types like `Integer` or `String`.
- The builder hides some complexity of what you're putting in and where
- It gives us a bit more flexibility in terms of changing the user facing configuration later I think as it's more decoupled

To be honest though, the builder already decouples things enough that we could do away with the `FixedPartitionDistribution` for example and just pass back a `Map<PartitionId, Map<MemberId, Integer>>` - however the nested map is unclear what the integer is, and it's not very extensible (though, you know, YAGNI and all...). We could compromise and just have the builder and nothing else.

Let me know what you think.

File: dist/src/main/config/broker.standalone.yaml.template [hunk 1]
```diff
@@ -518,6 +518,66 @@
...
+        #
+        # For each partition, you must specify its partition ID (ranging from 1 to
+        # (partitionCount - 1)), and the set of nodes which will manage it. This must be a subset
+        # of nodes whose IDs are valid within your cluster. This means with node IDs ranging from 0
+        # to the (clusterSize - 1), without any repeated IDs. This subset must have the same amount
+        # of nodes as the replication factor, otherwise the broker will not start.
+        #
+        # If you are using priority election, then you can also specify the priorities for each
+        # nodes. If you do not provide a priority, it will default to 1, which will still work but
+        # can have unintended consequences. Finally, the highest configured priority will become the
```
[the remark this answers is not in the document]
Comment (deepthidevaki) [reply dist/src/main/config/broker.standalone.yaml.template:552]: Also, I think we don't allow it to have same priority anymore. It fails the validation.

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/RaftPartitionGroupFactory.java [hunk 2]
```diff
@@ -91,4 +96,28 @@ RaftPartitionGroup buildRaftPartitionGroup(
...
+    }
+  }
+
+  private FixedPartitionDistributor buildFixedPartitionDistributor(final PartitioningCfg config) {
+    final var distributionBuilder =
+        new FixedPartitionDistributorBuilder(PartitionManagerImpl.GROUP_NAME);
+
+    for (final var partition : config.getFixed()) {
+      for (final var node : partition.getNodes()) {
+        distributionBuilder.assignMember(
```
Comment (deepthidevaki) [inline broker/src/main/java/io/camunda/zeebe/broker/partitioning/RaftPartitionGroupFactory.java:116]: The builder is nice :+1:

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
Thanks @npepinpe . Just have some minor comments.

The distribution builder looks good. But I think the `FixedPartitionDistributor` is unnecessary. It is just a wrapper around the map. You can remove it, but keep other abstractions.

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/distribution/FixedPartitionDistribution.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.broker.partitioning.distribution;
+
+import io.atomix.primitive.partition.PartitionId;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+final class FixedPartitionDistribution {
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply broker/src/main/java/io/camunda/zeebe/broker/partitioning/distribution/FixedPartitionDistribution.java:15]: In the initial approach, the map has a type: `Map<PartitionId, Map<MemberId, Integer>>`. Is this what you mean? Or do you refer to the wrapped map, `Map<PartitionId, Set<FixedDistributionMember>>`?

**Comment by npepinpe** [comment]
I applied your review hints, let me know what you think :+1:

**Review by deepthidevaki — APPROVED** [review]
:+1:

**Comment by npepinpe** [comment]
bors merge

**Comment by npepinpe** [comment]
bors retry

**Pull request merged**